server/src/main/java/io/crate/expression/scalar/ToCharFunction.java

```diff
@@ -0,0 +1,105 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import io.crate.data.Input;
+import io.crate.expression.symbol.Literal;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+import io.crate.types.IntervalType;
+import org.joda.time.Period;
+import org.joda.time.format.PeriodFormat;
+import org.joda.time.format.PeriodFormatter;
+import io.crate.types.TimestampType;
+
+import java.text.DecimalFormat;
+import java.text.NumberFormat;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.TimeZone;
+import java.util.stream.Stream;
+
+public class ToCharFunction extends Scalar<String, Object> {
+
+    public static final String NAME = "to_char";
+
+    public static void register(ScalarFunctionModule module) {
+        Stream.of(DataTypes.NUMERIC_PRIMITIVE_TYPES, List.of(DataTypes.TIMESTAMPZ, DataTypes.TIMESTAMP, DataTypes.INTERVAL))
+            .flatMap(Collection::stream)
+            .forEach(type -> {
+                module.register(
+                    Signature.scalar(
+                        NAME,
+                        type.getTypeSignature(),
+                        DataTypes.STRING.getTypeSignature(),
+                        DataTypes.STRING.getTypeSignature()
+                    ), ToCharFunction::new
+                );
+            });
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    public ToCharFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public String evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Object>... args) {
+        Literal expression = (Literal) args[0];
+        int dataType = expression.valueType().id();
+        if (TimestampType.ID_WITH_TZ == dataType) {
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern(String.valueOf(args[1].value()),
+                                                                      Locale.ENGLISH);
+            Long ts = DataTypes.TIMESTAMPZ.sanitizeValue(args[0].value());
```

Comment:
What do you think of using `BinaryScalar` and implementing keeping the evaluate logic a bit more separated? For example, see `CoordinateFunction` or `RoundFunction`.

Reply:
Aha! Yeah, I was just about to push a fixup where the evaluation logic was part of the registration step, rather than the evaluation step for every row. Thanks for the heads up :slightly_smiling_face: